[WIP] Smarter field maps

luisaleperez opened the pull request:
BEGINRELEASENOTES
- Updated Field maps readers, FieldMapBrBz and FieldMapXYZ, to automatically get from the root file the maps parameters
  - Coordinates ranges, step size and ordering

ENDRELEASENOTES

andresailer commented on detector/other/FieldMapBrBz.cpp:
```diff
@@ -52,71 +72,54 @@ void FieldMapBrBz::fieldComponents(const double* pos , double* globalField) {
   //field is supposed to be. Now we just assume it starts at 0/0/0, outside there is no field
-  if (not ( 0.0 <= x && x <= rhoMax &&
-	    0.0 <= y && y <= zMax ) ) {
+  if (not ( rhoMin <= r && r <= rhoMax &&
```
If you change these lines there will not be any bfield below rhoMin I.e. at the interaction point. This is not how it should be. See also the comment before this line.

luisaleperez replied:
Previous to these lines you have 

if(r < rhoMin) r = rhoMin;
if(z < zMin  ) z = zMin;

Which actully prevents that. If r  or z are smaller than rhoMin or zMin, respectively, then the values of r and z are update to be those minimum values.

andresailer replied:
Ok, right. But then you can also drop the comparison with rhoMin/zMin, because that is now always true.
And please change the comment above and move it to before setting r to rhoMin etc.

luisaleperez replied:
Done

andresailer commented on detector/other/FieldMapBrBz.cpp:
```diff
@@ -52,71 +72,54 @@ void FieldMapBrBz::fieldComponents(const double* pos , double* globalField) {
-  // Need addresses of these to pass to modf below.
-  // modf uses its second argument as an OUTPUT argument.
-  double xdindex, ydindex;
+  rBin = int((r - rhoMin)/rhoStep);
```
What was wrong with the previous way of calculating the local position and indices?

luisaleperez replied:
For me it is clearer this way.

andresailer replied:
Would be nice to have some comments for the different blocks of code, like it was before.

andresailer commented on detector/other/FieldMapBrBz.cpp:
```diff
@@ -150,44 +157,156 @@ void FieldMapBrBz::fillFieldMapFromTree(const std::string& filename, const std::
     throw std::runtime_error( error.str() );
   }
 
+  //Getting the tree name and branch variables names
```
Why is this needed? So that you can now use arbitrary order in the definition of the ntuple?
It was possible to define any field name before, as long as they had the same meaning. Now one can only use the same words except upper or lower case, can't one?

Before I could have a tree like
`myNewFieldMap:Radius:AlongTheDetector:RadialField:AlongTheDetectorField`
Now I can't.

luisaleperez replied:
You're right. Now you cannot define any tree variables names (coordinated and Bfield components), but you don't have to specify any given order. I did it this way because people usually use the same variables naming.

So, we can come back to the previous convention in which the names ordering is what is important, but this has the possibility of error by the user in specifying a wrong order. In this case the code can still work but with swapping of tree variables meaning.

Up to you to decide what is best.

andresailer commented on detector/other/FieldMapBrBz.cpp:
```diff
@@ -150,44 +157,156 @@ void FieldMapBrBz::fillFieldMapFromTree(const std::string& filename, const std::
+  for(int i=0;i<treeEntries;i++) {
+    tree->GetEntry(i);
+
+    if(i == 0) {
```
What if the tree is not ordered this way?

luisaleperez replied:
I suppose that the coordinates in the tree are scanned from lowest to highest values. If the tree doesn't follow it the code will crash with an error message.

We could as well verify a coordinate scanning order and compute the corresponding indexes accordingly. Do you think that it is worth doing?

andresailer replied:
Yes, true. This was already quite limited before.

luisaleperez replied:
I just did this. Now it is possible that a field map can contain one or mare variables ordered from high-to-low. I am testing this right now.

andresailer replied:
Great

andresailer commented on detector/other/FieldMapBrBz.cpp:
```diff
@@ -52,71 +72,54 @@ void FieldMapBrBz::fieldComponents(const double* pos , double* globalField) {
-    fv1.Bz * (1-xlocal) *    ylocal   +
-    fv2.Bz *    xlocal  * (1-ylocal)  +
-    fv3.Bz *    xlocal  *    ylocal   ;
+  field[0] = (1.0 - rd)*(1.0 - zd)*B_r0z0.Br + rd*(1.0 - zd)*B_r1z0.Br + (1.0 - rd)*zd*B_r0z1.Br + rd*zd*B_r1z1.Br;
```
I think the previous indentation/line breaks were much more readable.

luisaleperez replied:
I did the indentation/line breaks. I hope you like it now.

andresailer replied:
Well, you make spaces around the operator inside the parenthesis, but then you don't leave any space between the parenthesis and the next operand, which makes things harder to read, because the "optical" grouping is just opposite of the operator grouping.

andresailer commented on detector/other/FieldMapBrBz.cpp:
```diff
@@ -40,6 +40,26 @@ FieldMapBrBz::FieldMapBrBz() {
+  int zBin = int((z - zMin  )/zStep);
+  
+  int GlobalIndex = -1;
+  if(CoorsOrder == std::string("RZ"))      GlobalIndex = rBin + zBin*nRho;
```
String comparison in the look up for the bfield is a bad idea. This is _slow_!

luisaleperez replied:
OK. I will replace it by a int comparison.

luisaleperez replied:
Done

luisaleperez commented on detector/other/FieldMapBrBz.cpp:
```diff
@@ -40,6 +40,26 @@ FieldMapBrBz::FieldMapBrBz() {
 
 
 
+int FieldMapBrBz::GetBlobalIndex(double r, double z)
```
OK. I will replace the name of the function

luisaleperez replied:
Done

luisaleperez commented on detector/other/FieldMapBrBz.cpp:
```diff
@@ -52,71 +72,54 @@ void FieldMapBrBz::fieldComponents(const double* pos , double* globalField) {
   const double posPhi = atan2(pos[1], pos[0]);
 
   //Mokka defines these things as x and y and phi
```
I will update them.

luisaleperez replied:
Done

luisaleperez commented on detector/include/FieldMapXYZ.h:
```diff
@@ -17,8 +17,8 @@ class FieldMapXYZ: public DD4hep::Geometry::CartesianField::Object {
 
-  int nX, nY, nZ;
-  
+  std::string CoorsOrder;
```
CoorOrder is a string that tells you what the coordinates ordering in the tree. For example, in the case of FieldMapBrBz, it can have the values "RZ" or "ZR". In the first case, the coordinates are scanned first R and then Z in the tree. In the second case the scanning is on the other way around.

I changed this following a previous comment from you. Now CoorOrder is an integer which takes the values 1 or 2 depending if the ordering is "RZ" or "ZR", respectively.

luisaleperez commented:
I would like to make a final remark. The changes I implemented were motivated a lot of time wasted during a Beamstrahlung pre-production using a solenoidal field with the wrong ordering. So, I modified the code so that is protected from this bug.

luisaleperez commented:
I just modified the code to avoid errors in the specification of the tree and variables names. Now they are specified separately in the xml file. This increases the number of variables to specify but avoids ordering errors.

luisaleperez commented:
Check ILD/compact/ILD_o1_v05/ILD_o1_v05.xml for the new implementation.

andresailer commented on detector/other/FieldMapBrBz.cpp:
```diff
@@ -40,6 +40,26 @@ FieldMapBrBz::FieldMapBrBz() {
+  if(CoorsOrder == 1)      GlobalIndex = rBin + zBin*nRho;  //RZ coordinates ordering
+  else if(CoorsOrder == 2) GlobalIndex = zBin + rBin*nZ;    //ZR coordinates ordering
+  
+  if(GlobalIndex < 0 || GlobalIndex > nRho*nZ-1) {
```
Can this happen, given that `rMIn <= r < rMax` and `zMin <= z < zMax`?

luisaleperez replied:
Removed the check.

andresailer commented on detector/other/FieldMapBrBz.cpp:
```diff
@@ -51,72 +71,60 @@ void FieldMapBrBz::fieldComponents(const double* pos , double* globalField) {
-  // modf uses its second argument as an OUTPUT argument.
-  double xdindex, ydindex;
+  rBin = int((r - rhoMin)/rhoStep);
+  zBin = int((z - zMin)/zStep);
```
Here you calculate rBin and zBin, and use that to calculate r and z positions. Then inside getGlobalIndex you use r and z to calculate rBin and zBin. I think you could save a lot of these calculations and just do them once and increase rBin and zBin by one as needed to calculate the globalIndex for the fieldMap vector.

luisaleperez replied:
Now rBin and zBin are calculated just once. The getGlobalIndex function gets as inputs directly rBin and zBin.

andresailer commented:
Thanks for all your explanations and changes. I really think your "final remark" should have been in the comment when you opened the pull request.
The lookup of the bfield happens many times per step, so this shouldn't really be loaded with unnecessary calls and calculations.

luisaleperez commented:
I implemented your comments on FieldMapBrBz and also added similar improvements on FieldMapXYZ. I am currently testing the code. I will commit it very soon.

luisaleperez commented:
I implemented all the changes you suggested. Now the field map can be specified in any order of coordinates, and any coordinate can be specified from low-to-high or high-to-low.

gaede commented:
@luisaleperez can you please change all variable names to start with a lower case letter ?
Upper case is reserved for types or enums/constants.

gaede commented:
@luisaleperez why do you allow for different ordering in your internal field map ? This way you have to check every time a field value is created which ordering to use. Could you not always use the same fixed ordering internally and deal with this when reading/filling the field map ?

luisaleperez commented:
I just changed the variables names to always start in lower-case.

I also modified the code such that the field array is always filled in a given order. Then each access of the field performs less operations. I will check that it works properly next week.

andresailer commented on detector/other/FieldMapXYZ.cpp:
```diff
@@ -140,95 +181,168 @@ void FieldMapXYZ::fieldComponents(const double* pos , double* globalField) {
+  if(zStep < 0) {
     std::stringstream error;
-    error << "FieldMap[ERROR]: Tree " << strs[0] << " not found in file: " << filename;
+    error << "FieldMapBrBz[ERROR]: All z coordinates in n-tuple have the same value!!!";
```
Should be `FieldMapXYZ` here

andresailer commented:
Fine for me now, but we will wait for your tests next week.

luisaleperez commented:
I just finished the checks. The code works as expected. I think that we can proceed to pull the code.

gaede commented:
@luisaleperez why are you still use ordering parameter here:
https://github.com/luisaleperez/lcgeo/blob/c616c32f7c6a8d22511047e09159e5b534d32423/detector/other/FieldMapBrBz.cpp#L49-L54
?

luisaleperez commented:
Hello Frank,

This function is used only at the moment of the fill of the array with 
the field values inside the function fillFieldMapFromTree. This function 
is not called for calculating the field value at an arbitrary point 
inside the function fieldComponents.

Cheers,

   Alejandro


On 05/30/2017 09:55 AM, Frank Gaede wrote:
>
> @luisaleperez <https://github.com/luisaleperez> why are you still use 
> ordering parameter here:
> https://github.com/luisaleperez/lcgeo/blob/c616c32f7c6a8d22511047e09159e5b534d32423/detector/other/FieldMapBrBz.cpp#L49-L54
> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/iLCSoft/lcgeo/pull/103#issuecomment-304803014>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AYDhmttW8V1ZhQYtS2KxbSQe5J2A0tG1ks5r-8uBgaJpZM4Niige>.
>


-- 
   --------------------------------------------------------------------------------------
  | Luis Alejandro PÉREZ PÉREZ                                                           |
  | CRNS - Institut Pluridisciplinaire Hubert Curien (IPHC)                              |
  | Dept. de Recherches Subatomiques   phone  : +33 (0)3 88 10 68 04                     |
  | 23, rue du loess - BP 28           fax    : +33 (0)3 88 10 62 34                     |
  | F-67037- STRASBOURG cedex 2        e-mail : Luis_Alejandro.Perez_Perez@iphc.cnrs.fr  |
   --------------------------------------------------------------------------------------

gaede commented:
Ok, I see - will merge the PR then.
